icu4c/source/i18n/unicode/numberformatter.h

```diff
@@ -1127,6 +1125,61 @@ class U_I18N_API Scale : public UMemory {
 
 namespace impl {
 
+// Do not enclose entire Usage with #ifndef U_HIDE_INTERNAL_API, needed for a protected field
+/**
+ * Manages NumberFormatterSettings::usage()'s char* instance on the heap.
+ * @internal
+ */
+class U_I18N_API Usage : public UMemory {
+
+#ifndef U_HIDE_INTERNAL_API
+
+  public:
+    /** @internal */
+    Usage(const Usage& other);
+
+    /** @internal */
+    Usage& operator=(const Usage& other);
+
+    /** @internal */
+    Usage(Usage &&src) U_NOEXCEPT;
+
+    /** @internal */
+    Usage& operator=(Usage&& src) U_NOEXCEPT;
+
+    /** @internal */
+    ~Usage();
+
+    /** @internal */
+    int16_t length() const { return fLength; }
+
+    /** @internal
+     * Makes a copy of value.
```

Comment:
The main thing is we need to save the "usage" string in MacroProps instances. So this field is the main need, and in Java I'm quite confident it can indeed just be a plain String instance:

```
struct U_I18N_API MacroProps : public UMemory {
// [...]
    /** @internal */
    Usage usage;  // = Usage();  (no usage)
// [...]
```

In Java, a string is fine - we don't need to track error codes when we can throw exceptions, and we don't even need to copy strings (because they're immutable and garbage collected).

In C++, I think I added fError in order to be able to propagate errors happening while dealing with the Usage field. Looking at it now though, it looks like I've not done this correctly: I'm not checking for memory allocation errors, for example. I think it's mostly a bug that I'm not properly using fError - i.e. I still think it necessary to have a UErrorCode embedded with the Usage string (with the quartet of default copy/move construction in C++ being to blame).